### PR TITLE
Fix NameValueHeaderValue.CheckValueFormat

### PR DESCRIPTION
**Pull request by MarcoRossignoli**
Fixes https://github.com/dotnet/corefx/issues/30653 https://github.com/dotnet/runtime/issues/1504

RFC 7230 https://tools.ietf.org/html/rfc7230#section-3.2.6 states
```
Most HTTP header field values are defined using common syntax
components (token, quoted-string, and comment) separated by
whitespace or specific delimiting characters.  Delimiters are chosen
from the set of US-ASCII visual characters not allowed in a token
(DQUOTE and "(),/:;<=>?@[\]{}").
...
```

Current code try to do check searching for "token" length inside input, if the length returned is equal to 0 means that the start char of input is a not allowed token delimiter, in turn it tries to check if input is a valid quoted string. To do this check uses `GetValueLength` that returns a "value length" and if it's different from input length means that input is a "token" or an invalid quoted string.

This PR try to fix issue allowing "token", but I cannot reuse `GetValueLength` because simple "value length" is not enough to understand if it's valid or not. For instance a `GetValueLength` < inpunt.Length it's ok for a "token" value.

I added also new check for trailing/leading not allowed spaces, was "gratis" before.
Updated also a test becasue "x y" it's valid per RFC `separated by whitespace or specific delimiting characters`.

Some note: RFC talks also about "comments" that aren't covered by check.

cc: @scalablecory @davidsh

### CODE REVIEW

**Comment by MarcoRossignoli** [comment]
😢

File: src/libraries/System.Net.Http/src/System/Net/Http/Headers/NameValueHeaderValue.cs [hunk 1]
```diff
@@ -354,11 +354,27 @@ private static void CheckNameValueFormat(string name, string value)
...
-            // Either value is null/empty or a valid token/quoted string
-            if (!(string.IsNullOrEmpty(value) || (GetValueLength(value, 0) == value.Length)))
+            // Either value is null/empty or a valid token/quoted string https://tools.ietf.org/html/rfc7230#section-3.2.6
+            if (string.IsNullOrEmpty(value))
+            {
+                return;
+            }
+
+            // Trailing/leading space are not allowed
+            if (value[0] == ' ' || value[^1] == ' ')
```
Comment (jkotalik) [inline src/libraries/System.Net.Http/src/System/Net/Http/Headers/NameValueHeaderValue.cs]: Should we be checking for tabs here as well?

Comment (MarcoRossignoli) [reply src/libraries/System.Net.Http/src/System/Net/Http/Headers/NameValueHeaderValue.cs]: It's not super clear but I think so from https://tools.ietf.org/html/rfc7230#section-3.2.4
```
   OWS            = *( SP / HTAB )
   ...
   The field value does not
   include any leading or trailing whitespace: OWS occurring before the
   first non-whitespace octet of the field value or after the last
   non-whitespace octet of the field value ought to be excluded by
   parsers when extracting the field value from a header field.
```
I'll exclude HTAB and we'll see if it was expected(if covered in tests)

Comment (MarcoRossignoli) [reply src/libraries/System.Net.Http/src/System/Net/Http/Headers/NameValueHeaderValue.cs]: In past code throw for leading HTAB thanks to `HttpRuleParser.GetQuotedStringLength` but not for leading, so I added check for leading/trailing.

@jkotalik or @scalablecory can you please run outerloops?I cannot run CI yet.

Comment (scalablecory) [reply src/libraries/System.Net.Http/src/System/Net/Http/Headers/NameValueHeaderValue.cs]: having some chats with our team right now to find out why you can't run outerloop still :/

Comment (MarcoRossignoli) [reply src/libraries/System.Net.Http/src/System/Net/Http/Headers/NameValueHeaderValue.cs]: Thanks a lot Cory!

**Review by jkotalik — APPROVED** [review]
I'd get someone in NCL to confirm, but LGTM.

**Comment by scalablecory** [comment]
The change looks fine code wise but I need to take some time to give this a deeper look. This type is used in a lot of places.

**Comment by stephentoub** [comment]
> The change looks fine code wise but I need to take some time to give this a deeper look. This type is used in a lot of places.

@scalablecory, what are the next steps here?  How do we get this PR closed or merged?

File: src/libraries/System.Net.Http/src/System/Net/Http/Headers/NameValueHeaderValue.cs [hunk 2]
```diff
@@ -354,11 +354,27 @@ private static void CheckNameValueFormat(string name, string value)
...
-            // Either value is null/empty or a valid token/quoted string
-            if (!(string.IsNullOrEmpty(value) || (GetValueLength(value, 0) == value.Length)))
+            // Either value is null/empty or a valid token/quoted string https://tools.ietf.org/html/rfc7230#section-3.2.6
+            if (string.IsNullOrEmpty(value))
+            {
+                return;
+            }
+
+            // Trailing/leading space are not allowed
+            if (value[0] == ' ' || value[0] == '\t' || value[^1] == ' ' || value[^1] == '\t')
```
Comment (Youssef1313) [inline src/libraries/System.Net.Http/src/System/Net/Http/Headers/NameValueHeaderValue.cs:364]: What about other whitespace characters?

**Comment by stephentoub** [comment]
@scalablecory, @karelz, this has been sitting idle for a while.  What are the next steps to get this merged or closed?

**Comment by karelz** [comment]
@MarcoRossignoli looks like this dropped off our radar -- we will try hard to get back to it in next couple of weeks after we finish last 5.0 bugs in Networking space. Sorry for the XXXL delay :(

**Comment by MarcoRossignoli** [comment]
No worries Karel!

**Comment by stephentoub** [comment]
@karelz, can someone please follow-up on this?

**Comment by danmosemsft** [comment]
@karelz @scalablecory

**Review by scalablecory — APPROVED** [review]
lgtm. thanks for the PR and sorry this slipped by.

**Comment by danmosemsft** [comment]
@scalablecory I'd like to get this merged .. can you please confirm that none of the test failures are related? 

TestUserCredentialsPropertiesOnWindows  is now fixed (and also clearly unrelated) but there are some Http failures.

**Comment by scalablecory** [comment]
CI failures are unrelated.

**Pull request merged**